### PR TITLE
[DistillationTrainer refactor] Add the chunked JSD loss (unwired)

### PR DESCRIPTION
**Pull request by qgallouedec**
*Item 32 (Group E) — stacked on PR 31. Part of #6449.*

Adds the memory-efficient chunked JSD loss, modeled on SFT's `_chunked_cross_entropy_loss`. Still unwired: `compute_loss` runs the old full-logit `generalized_jsd_loss` path until item 33.

New (module-level, mirroring SFT's helpers):
- `_chunk(h_s, w_s, b_s, h_t, w_t, b_t, beta, valid)` — projects both student and teacher hidden states to vocab logits **inside** the gradient-checkpointed body (`.float()`), so only `(chunk, H)` is retained across the backward, never `(chunk, V)`. Computes the generalized JSD inline (same `F.kl_div` math as `_jsd_divergence`, no support mask) and returns per-chunk `(jsd_sum, entropy_sum)`, zeroing rows packed out of the valid prefix.
- `_chunked_divergence_loss(...)` — packs valid completion positions to the front (`argsort` on the mask, static-shape), rounds up to whole chunks, loops with `torch.utils.checkpoint.checkpoint(..., use_reentrant=False)`, reduces by `num_items_in_batch` (grad-accum-correct) or `mean` over local valid. Keeps the graph connected through the student hidden states + `lm_head` on a fully-masked micro-batch.
- `_CHUNKED_LM_HEAD_CHUNK_SIZE = 256`.

**Plan deviation (SFT parity):** the loss uses a module constant `_CHUNKED_LM_HEAD_CHUNK_SIZE`, not a `chunk_size` config field. SFT — the chunked-loss sibling — hardcodes the same constant and exposes no field, so adding one would be an addition with no sibling analog. Item 39's config work is unaffected.

Import: `maybe_gather_lm_head_ctx` from `trainer.utils` (SFT's ZeRO-3 gather context).

**Tests (the one deliberate loss-custody bespoke exception):** `TestChunkedDivergenceLoss` — exactness vs a naive full-vocab reference for β∈{0, 0.5, 1} × chunk sizes that divide / don't divide / exceed `n_valid`; β=1 == hand-rolled reverse KL; GKD parity (identity `lm_head`, sum reduction); masked positions ignored; grads flow at valid and are zero at masked; fully-masked batch keeps the graph. 16 tests. Verified: ruff clean; full suite 68 green.

---

> [!NOTE]
> **Low Risk**
> New helpers and tests only; training behavior is unchanged because the chunked loss is not connected to `compute_loss` yet.
> 
> **Overview**
> Adds a **memory-efficient chunked generalized JSD** path for distillation, modeled on SFT’s chunked cross-entropy, but **`compute_loss` still uses the existing full-logit `generalized_jsd_loss`** until a follow-up wires this in.
> 
> New module-level pieces in `distillation_trainer.py`: **`_chunk`** (student/teacher `lm_head` projection + JSD inside a gradient-checkpointed body, with ZeRO-3 **`maybe_gather_lm_head_ctx`**, teacher under `no_grad`, optional logit scale/softcap), **`_chunked_divergence_loss`** (pack valid completion tokens, loop chunks, FSDP2 `DTensor` handling, grad-accum via `num_items_in_batch`, zero-loss graph hook for fully masked batches), and **`_CHUNKED_LM_HEAD_CHUNK_SIZE = 256`**.
> 
> **`TestChunkedDivergenceLoss`** (16 tests) pins numeric parity vs a naive full-vocab reference, GKD, masking, gradients, and edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit c864c61dbf5fcbe2b9d73269cac8f9a24d614505. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/distillation/distillation_trainer.py [hunk 1]
```diff
@@ -162,6 +163,150 @@ def _jsd_divergence(student_log_probs, teacher_log_probs, beta, support_mask=Non
...
+            Teacher `lm_head` bias of shape `(V,)`, added to each chunk's logits when provided.
+
+    Returns:
+        `tuple[torch.Tensor, torch.Tensor, torch.Tensor]`: scalar loss, sum of per-token student entropy (in nats),
+        and number of valid completion positions — all over the local batch. Raw sums are returned so callers can
+        reduce correctly across ranks.
+    """
+    hidden_size = student_hidden_states.size(-1)
+    h_s = student_hidden_states.reshape(-1, hidden_size)
+    h_t = teacher_hidden_states.reshape(-1, hidden_size)
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py]: Fixed in bb7e824e. Each hidden-state tensor is now flattened with its own last dimension, so a teacher wider or narrower than the student projects correctly through its own lm_head. Added a test with mismatched teacher/student hidden sizes.

[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py]: https://github.com/huggingface/trl/pull/6526#discussion_r3642611770

File: trl/experimental/distillation/distillation_trainer.py [hunk 2]
```diff
@@ -162,6 +163,150 @@ def _jsd_divergence(student_log_probs, teacher_log_probs, beta, support_mask=Non
...
+    # Project both hidden states to vocab logits inside the checkpointed body so only `(chunk, H)` is retained across
+    # the backward, never `(chunk, V)`. ZeRO-3 shards the `lm_head`, so gather it tightly around each projection.
+    with maybe_gather_lm_head_ctx(w_s, b_s):
+        student_logits = h_s.float() @ w_s.float().t()
+        if b_s is not None:
+            student_logits = student_logits + b_s.float()
+    with maybe_gather_lm_head_ctx(w_t, b_t):
+        teacher_logits = h_t.float() @ w_t.float().t()
+        if b_t is not None:
+            teacher_logits = teacher_logits + b_t.float()
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py:188]: Fixed in bb7e824e. _chunk now applies per-model logit_scale (Cohere) and final_logit_softcapping (Gemma) before the softmax, matching each model's full forward, as SFT's chunked path does. They are independent for student and teacher since the two can be different families. Defaults are no-ops here; the wiring step sources them from each model's text_config via getattr, mirroring SFT. Added a test covering both transforms across beta.

File: trl/experimental/distillation/distillation_trainer.py [hunk 3]
```diff
@@ -162,6 +163,176 @@ def _jsd_divergence(student_log_probs, teacher_log_probs, beta, support_mask=Non
...
+            torch.stack([student_log_probs + torch.log1p(-beta_t), teacher_log_probs + torch.log(beta_t)]), dim=0
+        )
+        kl_teacher = F.kl_div(mixture_log_probs, teacher_log_probs, reduction="none", log_target=True)
+        kl_student = F.kl_div(mixture_log_probs, student_log_probs, reduction="none", log_target=True)
+        jsd = beta_t * kl_teacher + (1 - beta_t) * kl_student
+
+    # A chunk's tail may hold positions packed out of the valid prefix; zero those rows before summing.
+    per_token_jsd = jsd.sum(dim=-1) * valid
+    per_token_entropy = -(student_log_probs.exp() * student_log_probs).sum(dim=-1) * valid
+    return per_token_jsd.sum(), per_token_entropy.sum()
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py:215]: temperature is threaded through this loss and applied in the stacked item-33 PR (#6530, commit dfa8e99e), where compute_loss is actually wired to it. This function is unwired here, so temperature has no effect at this step regardless; it only matters once the loss is called, which is exactly where the fix lives. Same objective as the old generalized_jsd_loss at the default temperature=1.0.

File: trl/experimental/distillation/distillation_trainer.py [hunk 4]
```diff
@@ -162,6 +163,176 @@ def _jsd_divergence(student_log_probs, teacher_log_probs, beta, support_mask=Non
...
+        if b_s is not None:
+            student_logits = student_logits + b_s.float()
+    if s_scale != 1.0:
+        student_logits = student_logits * s_scale
+    if s_softcap is not None:
+        student_logits = s_softcap * torch.tanh(student_logits / s_softcap)
+    with maybe_gather_lm_head_ctx(w_t, b_t):
+        teacher_logits = h_t.float() @ w_t.float().t()
+        if b_t is not None:
+            teacher_logits = teacher_logits + b_t.float()
```
Comment (cmpatino) [inline trl/experimental/distillation/distillation_trainer.py]: Should this block enforce a no-grad computation of the logits?

AFAICT, we're keeping the gradients of `w_t` while calculating the teacher logits. Teacher gradients are never required in the distillation setting, so we might be using memory to store something we won't need at all during training.

I'd suggest adding a torch.no_grad() to the with block like below

```python
with maybe_gather_lm_head_ctx(w_t, b_t), torch.no_grad():
    teacher_logits = h_t.float() @ w_t.float().t()
    if b_t is not None:
        teacher_logits = teacher_logits + b_t.float()
```

We can also create a test that checks the grads for the teacher are `None` after running the backward pass. That way we ensure we're not wasting the memory gains from the chunked loss on unnecessary gradients.

Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py]: Good catch, done in e7742ae6. Wrapped the teacher projection in _chunk in torch.no_grad(). This matches how GRPO handles its ref model (whole ref forward under no_grad, params not frozen), and it's value-preserving.
